File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

function height(root){
     if(root==null){
        return 0
     }
     return (1+ Math.max(height(root.left), height(root.right)))
} 
var diameterOfBinaryTree = function(root) {
    
    if (root==null){
        return 0;
    }
    let leftdiameter = diameterOfBinaryTree(root.left)
    let rightdiameter = diameterOfBinaryTree(root.right);
    let diameter = height(root.left)+ height(root.right);

    return (Math.max(leftdiameter,rightdiameter,  diameter))
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* root, int &maxdiamter){
        if(root==NULL) return 0;
        int lh = helper(root->left,maxdiamter );
        int rh = helper(root->right,maxdiamter );
        maxdiamter= max(maxdiamter, lh+ rh);
        return (1+ max(lh,rh));
        
    }
    int diameterOfBinaryTree(TreeNode* root) {
         int maxdiamter = 0;
        int v = helper(root, maxdiamter);
        
        return maxdiamter;
        
    }
};